### ISSUE TITLE
Checkpoint transfer from convergence to production may not be sufficiently robust

### ISSUE THREAD

**Issue by eirrgang** [kassonlab/run_brer#10]
@carolinedav reported that production phase failed when simulations tried to start with zero-sized checkpoint files.

I'm not sure how this happened, but there are a few ways it could conceivably happen. We should figure out how to reproduce and examine log output (such as from line https://github.com/kassonlab/run_brer/blob/ae50d68468562a4589a4654ca2813cb5649f790c/run_brer/run_config.py#L135)

One set of issues could be related to the non-atomicity of file copies. Even leaving aside possible race conditions for the moment, there do not seem to be provisions for recovering from a partial copy of the checkpoint file after a failure at 
https://github.com/kassonlab/run_brer/blob/ae50d68468562a4589a4654ca2813cb5649f790c/run_brer/run_config.py#L153
in subsequent runs.

One way to make this section of code more robust would be to copy the checkpoint file to a temporary file in the target directory and then *move* the file (an atomic operation in all relevant environments) to its expected name only after the copy is successful. There are other measures we could take, but this is by far the simplest, and is a reasonable step to take regardless of anything else.

**Comment by eirrgang** [comment]
Observation: we shouldn't rely on just copying the checkpoint file to continue simulations. We should really generate a new TPR file using the old TPR file and the checkpoint file.

One thing to note is that different GROMACS versions have expressed different philosophies of how versatile or portable the checkpoint file is. I don't remember what the expected behavior is for GROMACS 2019, but it could be that different domain decomposition geometry for a relaunched simulation might render the checkpoint file invalid in poorly handled ways. Also note that the checkpoint file in this case is for a GROMACS 2019 simulation launched from a GROMACS 2018.1 TPR file, but restarted (production) from a new TPR file generated with GROMACS 2019 (due to the `end_time` argument used with the `md_sim` command).

**Comment by eirrgang** [comment]
I'm going to consider this issue resolved with the addition of an extra argument to the production phase `run()` command (accepting an optional TPR file from which to bootstrap the production phase).

If the recent patches do not prevent the symptom of zero-size checkpoint files, we can re-open.

Further issues of more robust transitions between phases should probably rely on upcoming gmxapi features in updated dependencies. We can track that in a separate issue, if needed.

As a workaround when the convergence phase checkpoint is not usable to start the production phase, the user can provide an alternative TPR file from which to bootstrap the production phase. Such a TPR file can be produced from the convergence checkpoint and the original inputs by providing the checkpoint file as the grompp trajectory input (`-t` option). However, there is no reliable way to obtain the grompp topology input from the required `run_brer` inputs (just TPR), so we can't provide this programmatically without changing the `run_brer` input requirements.

Since the main use case for the optional production-bootstrapping TPR is an incompatible checkpoint file, we need to override the default behavior of copying the checkpoint file into place when this kwarg is provided, and accept the blip in thermostat/barostat hysteresis caused by incomplete representation of simulator input state (through GROMACS 2021, at least).
